src/gennylib/include/gennylib/PhaseLoop.hpp

```diff
@@ -77,10 +78,36 @@ class IterationCompletionCheck final {
         }
     }
 
-    explicit IterationCompletionCheck(PhaseContext& phaseContext)
-        : IterationCompletionCheck(phaseContext.get<TimeSpec, false>("Duration"),
-                                   phaseContext.get<UIntSpec, false>("Repeat"),
-                                   phaseContext.isNop()) {}
+    explicit IterationChecker(PhaseContext& phaseContext)
+        : IterationChecker(phaseContext.get<TimeSpec, false>("Duration"),
+                           phaseContext.get<IntegerSpec, false>("Repeat"),
+                           phaseContext.isNop()) {
+        auto rateSpec = phaseContext.get<RateSpec, false>("Rate");
+        const auto rateLimiterName =
+            phaseContext.get<std::string, false>("RateLimiterName").value_or("defaultRateLimiter");
+
+        if (rateSpec) {
+            _rateLimiter =
+                phaseContext.workload().getRateLimiter(rateLimiterName, rateSpec.value());
+        }
+    }
+
+    bool shouldLimitRate(int64_t currentIteration) const {
+        // Only rate limit if the current iteration is a muliple of the burst size.
+        return _rateLimiter && (currentIteration % _rateLimiter->getBurstSize() == 0);
+    }
+
+    bool limitRate() const {
+        // Return if we're within the rate limit.
+        return _rateLimiter->consume();
+    }
+
+    void wait() const {
+        // Sleep for a little bit. This is not efficient if there's a large number of
+        // threads. If this is a problem, we can use exponential back off or wait
+        // based on the number of actors using this rate limiter.
+        std::this_thread::sleep_for(std::chrono::nanoseconds(_rateLimiter->getRate()));
```

Comment:
The idea of using sleep here is that the token bucket algorithm is lock free; using a condition variable would defeat the lock free-ness. Also, folly's own blocking version of consume() uses sleep: https://github.com/facebook/folly/blob/74ea53886c3333eda4eaf457d538a678ceaa5add/folly/TokenBucket.h#L266

I believe folly's design goal is to make the rate limit as light-weight as possible and leave the heavy-lifting (e.g. figuring how long to sleep) to callers. Using condition variables is probably going to make genny faster because we usually have many calling threads, but it would make the rate limiter itself slower.

Since using condition variables is going to make the rate limiter worse, I'm not sure it's something we should do. If we want to achieve the optimal performance for genny's immediate use cases, I think we should implement a separate "scheduler" and not use a rate limiter at all.

The distinction between the two is that the scheduler knowns all the information (rate spec and number of threads) in advance; all a thread has to do at run time is to sleep for a pre-computed duration (minus the length of the operation) This also guarantees that threads will go in some fixed order and never be starved.

A rate limiter is more versatile and allows different Y and X values for each call, but it also adds more runtime overhead.

cc @bcaimano in case you have thoughts